Maximum call stack size exceeded in HTML Attribute Mangler

# Bug report

- _WebMangler_: v0.1.19
- _WebMangler CLI_: v0.1.13
- _NodeJS_: v12.16.0
- _NPM_: v6.14.11

## Description

Using the [HTML Attribute Mangler](https://github.com/ericcornelissen/webmangler/blob/3def28489e308085811e5e5911c9a80990699683/packages/core/src/manglers/html-attributes.ts) on a large HTML source file (~4Mb) causes a `RangeError: Maximum call stack size exceeded` error when counting HTML attribute instances (as per the error log below).

It seems this issue is caused by the recent addition of [a check for `>` symbols in attribute values](https://github.com/ericcornelissen/webmangler/blob/3def28489e308085811e5e5911c9a80990699683/packages/core/src/languages/html/html-attributes.ts#L68-L72), as removing that prevents the error from being raised.

```
> webmangler --write -vv dist/

[DEBUG] reading configuration...
[DEBUG] configuration read
[DEBUG] reading files provided on the CLI...
[DEBUG] found 3 files
[DEBUG] mangling 3 files...
.../node_modules/webmangler/languages/utils/mangler-expression.class.js:101
        while ((match = regExp.exec(s)) !== null) {
                               ^

RangeError: Maximum call stack size exceeded
    at RegExp.exec (<anonymous>)
    at _ManglerExpression.exec (.../node_modules/webmangler/languages/utils/mangler-expression.class.js:101:32)
    at exec.next (<anonymous>)
    at .../node_modules/webmangler/engine.js:35:28
    at Array.forEach (<anonymous>)
    at .../node_modules/webmangler/engine.js:34:29
    at Array.forEach (<anonymous>)
    at .../node_modules/webmangler/engine.js:33:18
    at Array.forEach (<anonymous>)
    at countInstances (.../node_modules/webmangler/engine.js:31:11)
npm ERR! ...
```

## Suggestions

A potential solution is to omit the check in this Regular expression and create a separate one for this edge case. Another potential solution is to simplify the expression to have fewer capturing groups, e.g. the following replacement passes existing tests and prevents the above issue (at least for HTML files of ~4Mb):

```diff
-          (?<${GROUP_QUOTE}>"|')
-            (.(?!\\k<${GROUP_QUOTE}>))*
-            \\>
-            (.(?<!\\k<${GROUP_QUOTE}>))*
-          (\\k<${GROUP_QUOTE}>)
+          |
+          (?<="[^"]*)>(?=[^"]*")
+          |
+          (?<='[^']*)>(?=[^']*')
```

--- a/packages/core/CHANGELOG.md
+++ b/packages/core/CHANGELOG.md
@@ -8,7 +8,7 @@ Versioning].
 
 ## [Unreleased]
 
-- _No changes yet_
+- Fix _Maximum call stack size exceeded_ error by the HTML Attribute mangler.
 
 ## [0.1.9] - 2021-02-12
 

--- a/packages/core/src/languages/html/html-attributes.ts
+++ b/packages/core/src/languages/html/html-attributes.ts
@@ -4,9 +4,7 @@ import ManglerExpression from "../utils/mangler-expression.class";
 
 const GROUP_ALL = "all";
 const GROUP_ATTRIBUTE = "attribute";
-const GROUP_QUOTE = "quote";
 
-const SELECTOR_REQUIRED_BEFORE = "\\s";
 const SELECTOR_REQUIRED_AFTER = "\\s|\\=|\\>";
 
 /**
@@ -18,7 +16,7 @@ const SELECTOR_REQUIRED_AFTER = "\\s|\\=|\\>";
  */
 function getAttributeRegExp(pattern: string): RegExp {
   const expr = `
-    (?<=${SELECTOR_REQUIRED_BEFORE})
+    (?<=\\s|^)
     (?<${GROUP_ATTRIBUTE}>${pattern})
     (?=${SELECTOR_REQUIRED_AFTER})
   `.replace(/\s/g, "");
@@ -59,18 +57,14 @@ const expressions: ManglerExpression[] = [
   //  `<div (data-foo)="bar" (data-bar)="foo"></div>`
   new ManglerExpression(
     `
-      (?<=\\<\\s*[a-zA-Z]+(?=\\s))
+      (?<=\\<\\s*[a-zA-Z]+\\s+)
       (?<${GROUP_ALL}>
         (?:
-          [^\\>]
-          |
-          (?<${GROUP_QUOTE}>"|')
-            (.(?!\\k<${GROUP_QUOTE}>))*
-            \\>
-            (.(?<!\\k<${GROUP_QUOTE}>))*
-          (\\k<${GROUP_QUOTE}>)
+          (?:
+            [^>"']+
+            (?:\\=\\s*(?:"[^"]*"|'[^']*'))?
+          )\\s+
         )*
-        (?<=${SELECTOR_REQUIRED_BEFORE})
         (?<${GROUP_ATTRIBUTE}>%s)
         (?:${SELECTOR_REQUIRED_AFTER})
       )

--- a/packages/core/src/manglers/__tests__/html-attributes.test.ts
+++ b/packages/core/src/manglers/__tests__/html-attributes.test.ts
@@ -335,10 +335,10 @@ suite("HTML Attribute Mangler", function() {
             input: "<div data-foo=\"bar\"></div>",
             expected: "<div data-a=\"bar\"></div>",
           }),
-          {
+          ...varySpacing(">", {
             input: "<div data-foo></div>",
             expected: "<div data-a></div>",
-          },
+          }),
         ],
       },
       {
@@ -349,8 +349,12 @@ suite("HTML Attribute Mangler", function() {
             expected: "<div id=\"foo\" data-a=\"bar\"></div>",
           }),
           ...varyQuotes("html", {
-            input: "<div data-foo=\"bar\" id=\"foo\"></div>",
-            expected: "<div data-a=\"bar\" id=\"foo\"></div>",
+            input: "<div data-foo=\"foo\" class=\"bar\"></div>",
+            expected: "<div data-a=\"foo\" class=\"bar\"></div>",
+          }),
+          ...varyQuotes("html", {
+            input: "<div id=\"praise\" data-foo=\"the\" class=\"sun\"></div>",
+            expected: "<div id=\"praise\" data-a=\"the\" class=\"sun\"></div>",
           }),
           ...varyQuotes("html", {
             input: "<div id=\"foo\"><div data-foo=\"bar\"></div></div>",
@@ -465,6 +469,11 @@ suite("HTML Attribute Mangler", function() {
             input: "<div data-foo=\">\" data-bar=\"foo\"></div>",
             expected: "<div data-a=\">\" data-b=\"foo\"></div>",
             description: "closing `>` inside attribute values should be ignored",
+          },
+          {
+            input: "<div data-foo=\"bar\" data-foo=\"baz\"></div>",
+            expected: "<div data-a=\"bar\" data-a=\"baz\"></div>",
+            description: "Repeated attributes should all be mangled consistently",
           },
           {
             input: "<div class=\"data-foo\"></div>",

